pkg/test/ginkgo/retries.go

```diff
@@ -0,0 +1,282 @@
+	// Stop if we've hit max attempts
+	if attemptNumber > s.maxRetries {
+		logrus.Debugf("Stopping retry: max retries %d reached for %q", attemptNumber-1, originalFailure.name)
+		return false
+	}
+
+	// Skip retries for tests that exceed duration limit
+	if originalFailure.duration >= aggressiveMaximumTestDuration {
+		logrus.Debugf("Skipping retry: test duration %s exceeds limit %s", originalFailure.duration, aggressiveMaximumTestDuration)
+		return false
+	}
+
+	// Count current failures to avoid unnecessary retries
+	failureCount := 0
+	for _, attempt := range allAttempts {
+		if attempt.failed {
+			failureCount++
+		}
+	}
+
+	// Stop retrying if we've already reached the failure threshold - we know the test will fail
+	if failureCount >= s.failureThreshold {
+		logrus.Debugf("Stopping retry: failure threshold %d reached for %q", failureCount, originalFailure.name)
+		return false
+	}
+
+	// Early termination optimization: stop if it's mathematically impossible to reach failure threshold
+	// Calculate remaining attempts and check if we can still accumulate enough failures
+	// attemptNumber is the current attempt we're considering (1-based)
+	// maxRetries includes the original attempt, so total attempts possible = maxRetries
+	remainingAttempts := s.maxRetries - attemptNumber + 1
+	maxPossibleFailures := failureCount + remainingAttempts
+
+	// If even with all remaining attempts failing, we can't reach the threshold, stop early
+	if maxPossibleFailures < s.failureThreshold {
+		logrus.Debugf("Stopping retry: mathematically impossible to reach failure threshold %d for %q", s.failureThreshold, originalFailure.name)
+		return false
+	}
+
+	// In multi-retry mode, continue until we reach max attempts or failure threshold
+	return true
+}
+
+func (s *AggressiveRetryStrategy) DecideOutcome(testName string, attempts []*testCase) RetryOutcome {
+	failureCount := 0
+	skippedCount := 0
+
+	for _, attempt := range attempts {
+		if attempt.failed {
+			failureCount++
+		} else if attempt.skipped {
+			skippedCount++
+		}
+	}
+
+	// Only consider skipped if majority of attempts were skipped
+	if skippedCount > len(attempts)/2 {
+		return RetryOutcomeSkipped
+	}
+
+	if failureCount < s.failureThreshold {
+		return RetryOutcomeFlaky
+	}
+
+	return RetryOutcomeFail
+}
+
+type NoRetryStrategy struct{}
+
+func (s *NoRetryStrategy) Name() string { return "none" }
+func (s *NoRetryStrategy) ShouldAttemptRetries(failing []*testCase, suite *TestSuite) bool {
+	return false
+}
+func (s *NoRetryStrategy) GetMaxRetries(testCase *testCase) int { return 0 }
+func (s *NoRetryStrategy) ShouldContinue(testCase *testCase, allAttempts []*testCase, attemptNumber int) bool {
+	return false
+}
+func (s *NoRetryStrategy) DecideOutcome(testName string, attempts []*testCase) RetryOutcome {
+	panic("Unexpected call to NoRetryStrategy.DecideOutcome - this should never happen")
+}
+
+func getAvailableRetryStrategies() []string {
+	return []string{"once", "aggressive", "none"}
```

Comment:
WDYT of using a custom type here rather than a plain string? these are repeated a lot.